### PR TITLE
Allow new/delete operators to be called before DllMain

### PR DESCRIPTION
**Pull request by hekota**
Allow new/delete operators to be called before DllMain and allocator initialization in DxcInitThreadMalloc().

The operators can be called before DllMain from CRT libraries when static linking is enabled. If that happens, the new/delete operators will fallback to the standard allocator and use CoTaskMemAlloc/Free directly instead of CoGetMalloc, Alloc/Free & Release for better perf.

Enables fixing of [#5163](https://github.com/microsoft/DirectXShaderCompiler/issues/5163) and [#5178](https://github.com/microsoft/DirectXShaderCompiler/issues/5178).

### CODE REVIEW

File: lib/DxcSupport/dxcmem.cpp [hunk 1]
```diff
@@ -98,3 +98,39 @@ DxcThreadMalloc::DxcThreadMalloc(IMalloc *pMallocOrNull) throw() {
...
+#if defined(_WIN32) && !defined(DXC_DISABLE_ALLOCATOR_OVERRIDES)
+    ptr = CoTaskMemAlloc(size);
+#else
+    ptr = malloc(size);
+#endif
+  }
+  return ptr;
+}
+
+void DxcDelete(void *ptr) {
```
Comment (llvm-beanz) [inline lib/DxcSupport/dxcmem.cpp:121]: Is this safe? If an allocation gets allocated before `DllMain` and de-allocated after a custom allocator is enabled, what happens? Does that route the free to the custom allocator?

Comment (hekota) [reply lib/DxcSupport/dxcmem.cpp:121]: Alloc/free should always happen in the same context. The custom allocator is only enabled inside of compiler API calls. Every public API implementation uses `DxcThreadMalloc` local var to make sure the custom allocator is set to the thread-local storage for use and then re-set back to the previous one on exit.
@tex3d  - please correct me if I am wrong: The situation where something gets allocated by the default allocator (before or after `DllMain`) and de-allocated with a custom allocator should not happen.

**Review by llvm-beanz — APPROVED** [review]

File: lib/DxcSupport/dxcmem.cpp [hunk 2]
```diff
@@ -98,3 +98,39 @@ DxcThreadMalloc::DxcThreadMalloc(IMalloc *pMallocOrNull) throw() {
...
+  void *ptr;
+  IMalloc* iMalloc = DxcGetThreadMallocNoRef();
+  if (iMalloc != nullptr) {
+    ptr = iMalloc->Alloc(size);
+  } else {
+    // DxcGetThreadMallocNoRef() returning null means the operator is called before DllMain
+    // where the g_pDefaultMalloc is initialized, for example from CRT libraries when
+    // static linking is enabled. In that case fallback to the standard allocator
+    // and use CoTaskMemAlloc directly instead of CoGetMalloc, Alloc & Release for better perf.
+#if defined(_WIN32) && !defined(DXC_DISABLE_ALLOCATOR_OVERRIDES)
```
Comment (pow2clk) [inline lib/DxcSupport/dxcmem.cpp:112]: Not strictly necessary for non-win platforms as they already have `CoTaskMemAlloc` and `CoTaskMemFree` defined to malloc and free in WinAdapter.h:
https://github.com/microsoft/DirectXShaderCompiler/blob/93555aa911171097d72036df15b2e3c5c0c3d893/include/dxc/WinAdapter.h#L46

I'm unsure if this is needed for `DXC_DISABLE_ALLOCATOR_OVERRIDES`. #5296 didn't change the usages of `CoTaskMem*` in `dxcompilerobj.cpp`: 
https://github.com/microsoft/DirectXShaderCompiler/blob/93555aa911171097d72036df15b2e3c5c0c3d893/tools/clang/tools/dxcompiler/dxcompilerobj.cpp#L1510C40-L1510C40

**Review by pow2clk — APPROVED** [review]
I think it's harmless, but I'm unsure if the explicit free/malloc usage is actually needed

**Pull request merged**